File: app/src/main/java/cz/vitskalicky/lepsirozvrh/bakaAPI/rozvrh/rozvrh3/RozvrhConverter.java
package cz.vitskalicky.lepsirozvrh.bakaAPI.rozvrh.rozvrh3;

import android.content.Context;

import org.joda.time.LocalDate;
import org.joda.time.format.DateTimeFormat;

import java.util.ArrayList;
import java.util.HashMap;

import cz.vitskalicky.lepsirozvrh.R;
import cz.vitskalicky.lepsirozvrh.items.Rozvrh;
import cz.vitskalicky.lepsirozvrh.items.RozvrhDen;
import cz.vitskalicky.lepsirozvrh.items.RozvrhHodina;
import cz.vitskalicky.lepsirozvrh.items.RozvrhHodinaCaption;
import cz.vitskalicky.lepsirozvrh.items.RozvrhRoot;

public class RozvrhConverter {
    public static RozvrhRoot convert(Rozvrh3 rozvrh3, boolean perm, Context context){
        Rozvrh.MutableRozvrh rozvrh = new Rozvrh.MutableRozvrh();
        rozvrh.setTyp(perm ? "perm" : "akt");

        rozvrh.setNazevcyklu(rozvrh3.cycles.length > 0 ? rozvrh3.cycles[0].name : "");
        rozvrh.setZkratkacyklu(rozvrh3.cycles.length > 0 ? rozvrh3.cycles[0].abbrev : "");

        ArrayList<RozvrhHodinaCaption> captions = new ArrayList<>();
        for (Hour3 item :rozvrh3.hours) {
            RozvrhHodinaCaption nev = new RozvrhHodinaCaption();
            nev.setCaption(item.caption);
            nev.setBegintime(item.beginTime);
            nev.setEndtime(item.endTime);
            captions.add(nev);
        }
        rozvrh.setHodiny(captions);

        HashMap<String, Hour3> hours = new HashMap<>();
        for (Hour3 item :rozvrh3.hours) {
            hours.put(item.id + "", item);
        }
        HashMap<String, Class3> classes = new HashMap<>();
        for (Class3 item :rozvrh3.classes) {
            classes.put(item.id, item);
        }
        HashMap<String, Group3> groups = new HashMap<>();
        for (Group3 item :rozvrh3.groups) {
            groups.put(item.id, item);
        }
        HashMap<String, Subject3> subjects = new HashMap<>();
        for (Subject3 item :rozvrh3.subjects) {
            subjects.put(item.id, item);
        }
        HashMap<String, Teacher3> teachers = new HashMap<>();
        for (Teacher3 item :rozvrh3.teachers) {
            teachers.put(item.id, item);
        }
        HashMap<String, Room3> rooms = new HashMap<>();
        for (Room3 item :rozvrh3.rooms) {
            rooms.put(item.id, item);
        }
        HashMap<String, Cycle3> cycles = new HashMap<>();
        for (Cycle3 item :rozvrh3.cycles) {
            cycles.put(item.id, item);
        }

        String[] daysOfWeek = {
                context.getString(R.string.monday),
                context.getString(R.string.tuesday),
                context.getString(R.string.wednesday),
                context.getString(R.string.thursday),
                context.getString(R.string.friday),
                context.getString(R.string.saturday),
                context.getString(R.string.sunday)
        };

        ArrayList<RozvrhDen> days = new ArrayList<>();
        for (Day3 item :rozvrh3.days) {
            RozvrhDen newDen = new RozvrhDen();
            if (!perm) {
                LocalDate date = DateTimeFormat.forPattern("yyyy-MM-dd'T'HH:mm:ssZZ").parseLocalDate(item.date);
                newDen.setDatum(date.toString(RozvrhDen.DATE_FORMATTER));
            }else {
                newDen.setDatum("");
            }
            newDen.setZkratka(daysOfWeek[item.dayOfWeek - 1]);

            ArrayList<RozvrhHodina> lessons = new ArrayList<>();
            for (Atom3 atom :item.atoms) {
                RozvrhHodina newHodina = new RozvrhHodina();
                Hour3 caption = hours.get(atom.hourId);
                newHodina.setCaption(caption == null ? "" : caption.caption);
                newHodina.setTyp(atom.subjectId == null ? "X" : "H");

                Subject3 subject = atom.subjectId != null ? subjects.get(atom.subjectId) : null;
                newHodina.setZkrpr(subject != null ? subject.abbrev : "");
                newHodina.setPr(subject != null ? subject.name : "");

                Teacher3 teacher = atom.teacherId != null ? teachers.get(atom.teacherId) : null;
                newHodina.setZkruc(teacher != null ? teacher.abbrev : "");
                newHodina.setUc(teacher != null ? teacher.name : "");

                Room3 room = atom.roomId != null ? rooms.get(atom.roomId) : null;
                newHodina.setZkrmist(room != null ? room.abbrev : "");
                newHodina.setMist(room != null ? room.name : "");

                newHodina.setAbs(""); //ignored
                newHodina.setTema(atom.theme);

                StringBuilder zkrSkupSb = new StringBuilder();
                StringBuilder skupSb = new StringBuilder();
                boolean first = true;
                if (atom.groupIds != null){
                    for (int i = 0; i < atom.groupIds.length; i++) {
                        Group3 group = groups.get(atom.groupIds[i]);
                        if (group != null){
                            if (!first){
                                zkrSkupSb.append(", ");
                                skupSb.append(", ");
                            }
                            first = false;
                            zkrSkupSb.append(group.abbrev);
                            skupSb.append(group.name);
                        }
                    }
                }
                newHodina.setZkrskup(zkrSkupSb.toString());
                newHodina.setSkup(skupSb.toString());

                StringBuilder cycleSb = new StringBuilder();
                if (atom.cycleIds != null) {
                    for (int i = 0; i < atom.cycleIds.length; i++) {
                        Cycle3 cycle = cycles.get(atom.cycleIds[i]);
                        cycleSb.append(cycle != null ? cycle.abbrev : "");
                    }
                }
                newHodina.setCycle(cycleSb.toString());

                newHodina.setChng(atom.change == null ? "" : atom.change.description);
                if (atom.change != null){
                    if (atom.change.changeType.equals("Canceled") || atom.change.changeType.equals("Removed")){
                        newHodina.setTyp("X");
                    }
                    if (atom.change.typeAbbrev != null){
                        newHodina.setTyp("A");
                        newHodina.setZkrpr(atom.change.typeAbbrev);
                        newHodina.setPr(atom.change.typeName);
                    }

                }

                newHodina.setUkolodevzdat(atom.homeworkIds.length > 0 ? Integer.toString(atom.homeworkIds.length) : "");

                newHodina.commit();

                lessons.add(newHodina);
            }
            newDen.setHodiny(lessons);
            days.add(newDen);
        }

        rozvrh.setDny(days);
        rozvrh.onCommit();
        RozvrhRoot rozvrhRoot = new RozvrhRoot();
        rozvrhRoot.setRozvrh(rozvrh);

        return rozvrhRoot;
    }
}

File: app/src/main/java/cz/vitskalicky/lepsirozvrh/bakaAPI/rozvrh/rozvrh3/RozvrhConverter.kt
package cz.vitskalicky.lepsirozvrh.bakaAPI.rozvrh.rozvrh3

import cz.vitskalicky.lepsirozvrh.Utils
import cz.vitskalicky.lepsirozvrh.model.relations.BlockRelated
import cz.vitskalicky.lepsirozvrh.model.relations.DayRelated
import cz.vitskalicky.lepsirozvrh.model.relations.RozvrhRelated
import cz.vitskalicky.lepsirozvrh.model.rozvrh.*
import org.joda.time.DateTime
import org.joda.time.LocalDate
import org.joda.time.LocalTime
import org.joda.time.format.DateTimeFormat
import kotlin.collections.ArrayList
import kotlin.collections.HashMap

object RozvrhConverter {
    @Throws(RozvrhConversionException::class)
    fun convert(rozvrh3: Rozvrh3, date: LocalDate?): RozvrhRelated{
        //todo perform further testing after creating a testing server
        val rozvrh3 = remove0thCaptionIfUnnecessary(rozvrh3)

        val monday : LocalDate = date?.let { Utils.getWeekMonday(date) } ?: Rozvrh.PERM
        val cycle: RozvrhCycle? = if (date == null){
                null
            }else{
                if (rozvrh3.cycles.isEmpty()){
                    RozvrhCycle("","","")
                }else{
                    val c3 = rozvrh3.cycles[0]
                    RozvrhCycle(c3.id, c3.name, c3.abbrev)
                }
            }

        val rozvrh = Rozvrh(monday, DateTime.now(), monday == Rozvrh.PERM, cycle)

        //caption3 id and corresponding RozvrhCaption
        val captionsUnsorted = ArrayList<Pair<String,RozvrhCaption>>()
        for (value in rozvrh3.hours.withIndex()) {
            val item = value.value
            val nev = RozvrhCaption(
                    monday,
                    monday.toString() + "-" + (item.id.hashCode() * item.beginTime.hashCode()).hashCode().toString(16),
                    item.caption,
                    LocalTime.parse(item.beginTime),
                    LocalTime.parse(item.endTime),
                    value.index
            )
            captionsUnsorted.add(Pair(item.id.toString(), nev))
        }

        //to be extra sure, we sort the caption ascending by begin time to make sure it has the right index
        captionsUnsorted.sortWith( compareBy { it.second.beginTime } )
        //here we have the RozvrhCaptions
        val captionsMap = HashMap<String, RozvrhCaption>()
        captionsUnsorted.forEachIndexed { index, pair -> captionsMap[pair.first] = pair.second.copy(index = index) }
        val captions: List<RozvrhCaption> = captionsUnsorted.mapIndexed { index, pair -> pair.second.copy(index = index) }

        val hours = HashMap<String, Hour3>()
        for (item in rozvrh3.hours) {
            hours[item.id.toString() + ""] = item
        }
        val classes = HashMap<String, Class3>()
        for (item in rozvrh3.classes) {
            classes[item.id] = item
        }
        val groups = HashMap<String, Group3>()
        for (item in rozvrh3.groups) {
            groups[item.id] = item
        }
        val subjects = HashMap<String, Subject3>()
        for (item in rozvrh3.subjects) {
            subjects[item.id] = item
        }
        val teachers = HashMap<String, Teacher3>()
        for (item in rozvrh3.teachers) {
            teachers[item.id] = item
        }
        val rooms = HashMap<String, Room3>()
        for (item in rozvrh3.rooms) {
            rooms[item.id] = item
        }
        val cycles = HashMap<String, Cycle3>()
        for (item in rozvrh3.cycles) {
            cycles[item.id] = item
        }

        val days = ArrayList<DayRelated>()

        for (item in rozvrh3.days) {

            var dayDate : LocalDate = if (monday != Rozvrh.PERM) {
                DateTimeFormat.forPattern("yyyy-MM-dd'T'HH:mm:ssZZ").parseLocalDate(item.date)
            }else{
                Rozvrh.PERM.plusDays(item.dayOfWeek - 1)
            }
            val event: String? = if (item.dayDescription?.isNotBlank() == true){
                    item.dayDescription
                }else{
                    null
                }

            val day = RozvrhDay(dayDate, monday, event)

            val lessons = Array<ArrayList<RozvrhLesson>>(captions.size) { ArrayList() }
            for (atom in item.atoms) {
                val caption: RozvrhCaption = captionsMap[atom.hourId] ?:
                //report problem
                throw RozvrhConversionException("Failed to parse Rozvrh3 to Rozvrh: Could not find a caption for an atom: searched for '${atom.hourId}' available caption ids: ${captionsMap.keys}")

                val captionId: String = caption.id

                var subjectName: String = ""
                var subjectAbbrev: String = ""

                atom.subjectId?.let { subjects[it] }?.let{
                    subjectName = it.name ?: ""
                    subjectAbbrev = it.abbrev ?: ""
                }

                var teacherName: String = ""
                var teacherAbbrev: String = ""

                atom.teacherId?.let { teachers[it] }?.let {
                    teacherName = it.name
                    teacherAbbrev = it.abbrev
                }

                var roomName: String = ""
                var roomAbbrev: String = ""

                atom.roomId?.let { rooms[it] }?.let {
                    roomName = it.name
                    roomAbbrev = it.abbrev
                }

                val theme = atom.theme ?: ""

                var changeType: Int = RozvrhLesson.NO_CHANGE
                var chngDesc: String? = null
                if (atom.change != null) {
                    chngDesc = atom.change.description
                    changeType = RozvrhLesson.CHANGED
                    if (!atom.change.typeAbbrev.isNullOrBlank()) {
                        changeType = RozvrhLesson.CANCELLED
                        subjectAbbrev = atom.change.typeAbbrev
                        subjectName = atom.change.typeName ?: ""
                    }
                }

                val lessonGroups = ArrayList<RozvrhGroup>()
                atom.groupIds?.forEach {
                    groups[it]?.let{
                        lessonGroups.add(RozvrhGroup(it.id, it.name, it.abbrev))
                    }
                }

                val lessonCycles = ArrayList<RozvrhCycle>()
                atom.cycleIds?.forEach {
                    cycles[it]?.let {
                        lessonCycles.add(RozvrhCycle(it.id, it.name, it.abbrev))
                    }
                }

                val homeworkIds = ArrayList<String>()
                atom.homeworkIds?.map {
                    if (it.length > 3 && atom.groupIds != null){
                        val id = it.substring(2, 4)
                        for (grp in atom.groupIds) {
                            if (grp == id) {
                                homeworkIds.add(it)
                                break
                            }
                        }
                    }
                }

                lessons[caption.index].add(RozvrhLesson(
                        subjectName,
                        subjectAbbrev,
                        teacherName,
                        teacherAbbrev,
                        roomName,
                        roomAbbrev,
                        lessonGroups,
                        lessonCycles,
                        homeworkIds,
                        theme,
                        changeType,
                        chngDesc
                ))
            }

            val blocks = ArrayList<BlockRelated>(captions.size)

            for (capt in captions){
                blocks.add(BlockRelated(
                        RozvrhBlock(
                            dayDate,
                            capt.id,
                            lessons[capt.index]
                        ),
                        capt
                    )
                )
            }
            days.add(DayRelated(day, blocks))
        }
        
        return RozvrhRelated(rozvrh, captions, days)
    }

    /**
     * Romeves 0th caption if present and if unnecessary and return a modified copy.
     */
    fun remove0thCaptionIfUnnecessary(rozvrh3: Rozvrh3): Rozvrh3{
        val zeroCaptions = rozvrh3.hours.filter { it.caption.trim() == "0" }
        if (zeroCaptions.size != 1 ){
            return rozvrh3
        }
        val zeroCaption = zeroCaptions[0]

        var isEmpty: Boolean = true;
        rozvrh3.days.forEach {
            isEmpty = isEmpty && it.atoms.none { it.hourId == zeroCaption.id.toString() }
        }
        if (!isEmpty){
            return rozvrh3
        }
        return rozvrh3.copy(hours = rozvrh3.hours.toMutableList().apply { removeAll{ it.id == zeroCaption.id }})
    }

    class RozvrhConversionException(message: String): RuntimeException(message)
}